pkg/controller/podautoscaler/horizontal.go

```diff
@@ -428,45 +428,11 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 			rescaleReason = "All metrics below target"
 		}
 
-		// Do not upscale too much to prevent incorrect rapid increase of the number of master replicas caused by
-		// bogus CPU usage report from heapster/kubelet (like in issue #32304).
-		scaleUpLimit := calculateScaleUpLimit(currentReplicas)
-
-		switch {
-		case desiredReplicas > scaleUpLimit:
-			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "ScaleUpLimit", "the desired replica count is increasing faster than the maximum scale rate")
-			desiredReplicas = scaleUpLimit
-
-			// Ensure that even if the scaleUpLimit is greater
-			// than the maximum number of replicas, we only
-			// set the max number of replicas as desired.
-			if scaleUpLimit > hpa.Spec.MaxReplicas {
-				setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooManyReplicas", "the desired replica count was more than the maximum replica count")
-				desiredReplicas = hpa.Spec.MaxReplicas
-			}
+		desiredReplicas, err = a.normalizeDesiredReplicas(hpa, currentReplicas, desiredReplicas)
 
-		case hpa.Spec.MinReplicas != nil && desiredReplicas < *hpa.Spec.MinReplicas:
-			// make sure we aren't below our minimum
-			var statusMsg string
-			if desiredReplicas == 0 {
-				statusMsg = "the desired replica count was zero"
-			} else {
-				statusMsg = "the desired replica count was less than the minimum replica count"
-			}
-
-			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooFewReplicas", statusMsg)
-			desiredReplicas = *hpa.Spec.MinReplicas
-		case desiredReplicas == 0:
-			//  never scale down to 0, reserved for disabling autoscaling
-			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooFewReplicas", "the desired replica count was zero")
-			desiredReplicas = 1
-		case desiredReplicas > hpa.Spec.MaxReplicas:
-			// make sure we aren't above our maximum
-			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooManyReplicas", "the desired replica count was more than the maximum replica count")
-			desiredReplicas = hpa.Spec.MaxReplicas
-		default:
-			// mark that we're within acceptible limits
-			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionFalse, "DesiredWithinRange", "the desired replica count is within the acceptible range")
+		if err != nil {
+			a.eventRecorder.Eventf(hpa, v1.EventTypeWarning, "FailedNormalizeDesiredReplicas", err.Error())
```

Comment:
I think this event will be hard to understand for anyone who doesn't actually know this code (ie. doesn't know that a function `normalizeDesiredReplicas` exists).

Reply:
Agreed! I think @DirectXMan12 suggestion was to delete this `err` case entirely, and I'm inclined to agree with him.